File: leetcode/problems/139/139.cc
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp;
        dp.resize(s.length());
        string part = "";
        unordered_set<string> d;
        for(auto s : wordDict){
            d.insert(s);
        }
        for(int i = 0; i < s.length(); i++){
            part += s[i];
            if(d.find(part) != d.end()){
                part = "";
                dp[i] = true; 
            }
        }
        return dp[s.length()-1];
    }
};

File: leetcode/problems/139/139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        N = len(s)
        
        dp = [False] * (N+1)
        words = set(wordDict)
        
        dp[0] = True
        
        for i in range(0, N+1):
            for j in range(i):
                substr = s[j:i]

                if dp[j] and substr in words:
                    dp[i] = True
                    break

        return dp[len(s)]